File: api/src/main/java/dev/obscuria/fragmentum/network/PayloadCodec.java
package dev.obscuria.fragmentum.network;

import com.mojang.serialization.Codec;
import net.minecraft.core.RegistryAccess;
import net.minecraft.network.FriendlyByteBuf;
import org.apache.commons.lang3.NotImplementedException;

import java.util.function.Supplier;

@SuppressWarnings("unused")
public interface PayloadCodec<T> {

    static <T> PayloadCodec<T> registryFriendly(Codec<T> codec, Supplier<RegistryAccess> registryAccess) {
        throw new NotImplementedException();
    }

    void write(FriendlyByteBuf buf, T value);

    T read(FriendlyByteBuf buf);
}

File: api/src/main/kotlin/dev/obscuria/fragmentum/network/PayloadCodec.kt
@file:Suppress("unused", "UNUSED_PARAMETER", "RedundantNullableReturnType", "UnusedReceiverParameter")

package dev.obscuria.fragmentum.network

import com.mojang.serialization.Codec
import net.minecraft.core.RegistryAccess
import net.minecraft.network.FriendlyByteBuf

interface PayloadCodec<T> {

    fun write(buf: FriendlyByteBuf, value: T & Any)

    fun read(buf: FriendlyByteBuf): T

    companion object {

        fun <T> registryFriendly(codec: Codec<T>, registry: () -> RegistryAccess): PayloadCodec<T> = error("API stub")
    }
}